pyriemann_qiskit/classification.py

```diff
@@ -743,3 +750,244 @@ def predict(self, X):
+    corresponding to each class. There is no training. Calculating a distance
+    to a hull is an optimization problem and it is calculated for each testing
+    sample (SPD matrix) and each hull/class. The minimal distance defines the
+    predicted class.
+
+    Notes
+    -----
+    .. versionadded:: 0.2.0
+    """
+
+    def __init__(self, n_jobs=6, n_hulls=3, n_samples_per_hull=10):
+        """Init."""
+        self.n_jobs = n_jobs
+        self.n_samples_per_hull = n_samples_per_hull
+        self.n_hulls = n_hulls
+        self.matrices_per_class_ = {}
+        self.debug = False
+
+    def fit(self, X, y):
+        """Fit (store the training data).
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_matrices, n_channels, n_channels)
+            Set of SPD matrices.
+        y : ndarray, shape (n_matrices,)
+            Labels for each matrix.
+        sample_weight : None
+            Not used, here for compatibility with sklearn API.
+
+        Returns
+        -------
+        self : NearestNeighbor instance
+            The NearestNeighbor instance.
+        """
+
+        if self.debug:
+            print("Start NCH Train")
+        self.classes_ = np.unique(y)
+
+        for c in self.classes_:
+            self.matrices_per_class_[c] = X[y == c, :, :]
+
+        if self.debug:
+            print("Samples per class:")
+            for c in self.classes_:
+                print("Class: ", c, " Count: ", self.matrices_per_class_[c].shape[0])
+
+            print("End NCH Train")
+
+    def _predict_distances(self, X):
+        """Helper to predict the distance. Equivalent to transform."""
+        dist = []
+
+        if self.debug:
+            print("Total test samples:", X.shape[0])
+
+        parallel = self.n_jobs > 1
+
+        if self.debug:
+            if parallel:
+                print("Running in parallel")
+            else:
+                print("Not running in parallel")
+
+        if parallel:
+            dist = Parallel(n_jobs=self.n_jobs)(
+                delayed(self._process_sample)(test_sample) for test_sample in X
+            )
+
+        else:
+            for test_sample in X:
+                dist_sample = self._process_sample(test_sample)
+                dist.append(dist_sample)
+
+        return dist
+
+    def _process_sample(self, test_sample):
+        distances = []
+
+        for c in self.classes_:
+            total_distance = 0
+
+            # using multiple hulls
+            for i in range(0, self.n_hulls):
+                random_samples = random.sample(
+                    range(self.matrices_per_class_[c].shape[0]),
+                    k=self.n_samples_per_hull,
+                )
+
+                hull_data = self.matrices_per_class_[c][random_samples, :, :]
+
+                distance = qdistance_logeuclid_to_convex_hull(hull_data, test_sample)
+                total_distance = total_distance + distance
```

Comment:
OK so this is another strategy. So instead of summing the distances you choose the lowest distance out of all the hulls and this one gives you the predicted class?

Reply:
Yes, but this approach is still not satisfactory, because we could randomly select matrices very far away.

So, we can compute a single distance to convex hull, but on the 10 closest matrices (`n_closest_samples`=10).
First, we compute distances between test matrix and matrices of the class.
Second, we keep the 10 closest matrices.
Finally, we compute qdistance_logeuclid_to_convex_hull on these matrices.

Reply:
The 10 closests matrices that don t intersect?

Reply:
> Yes, but this approach is still not satisfactory, because we could randomly select matrices very far away, giving bad results.

I agree. To alleviate this problem I was trying to use many hulls like 15 for example.

Reply:
Yes. My suggestion is "one hull with the 10 closest matrices".

Reply:
So this will be implemented in #254